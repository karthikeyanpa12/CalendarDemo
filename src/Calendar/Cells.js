import React from 'react';
import dateFns from "date-fns";

export default function Cells(props) {
    const { currentMonth, selectedDate, reminders, onDateClick } = props;
    const monthStart = dateFns.startOfMonth(currentMonth);
    const monthEnd = dateFns.endOfMonth(monthStart);
    const startDate = dateFns.startOfWeek(monthStart, { weekStartsOn: 1 });
    const endDate = dateFns.endOfWeek(monthEnd);
    const dateFormat = "D", rows = []
    let days = [], day = startDate, formattedDate = "";

    while (day <= endDate) {
        for (let i = 0; i < 7; i++) {
            formattedDate = dateFns.format(day, dateFormat);
            const selectedDate = reminders.filter(obj => obj.date === dateFns.format(day, "DD/MM/YYYY"));
            const selected = selectedDate.length > 0 ? 'selected' : ''
            const cloneDay = day;
            days.push(
                <div className={`col cell ${selected}`}
                    key={day} onClick={() => onDateClick(dateFns.parse(cloneDay))}>
                    <span className="number">{formattedDate}</span>
                    <span className="message">{selectedDate && selectedDate[0] && selectedDate[0].message}</span>
                </div>);
            day = dateFns.addDays(day, 1);
        }
        rows.push(<div className="row" key={day}>  {days}</div>);
        days = [];
    }
    return <div className="body">{rows}</div>;
}