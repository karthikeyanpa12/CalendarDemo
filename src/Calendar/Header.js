import React from 'react';
import dateFns from "date-fns";

export default function header(props) {
    const dateFormat = "MMMM YYYY";
    const { prevMonth, currentMonth, nextMonth } = props;
    return (
        <div className="header row flex-middle">
            <div className="col col-start">
                <div className="icon" onClick={prevMonth}> Previous</div>
            </div>
            <div className="col col-center">
                <span>{dateFns.format(currentMonth, dateFormat)}</span>
            </div>
            <div className="col col-end" onClick={nextMonth}>
                <div className="icon">Next</div>
            </div>
        </div>
    );
}