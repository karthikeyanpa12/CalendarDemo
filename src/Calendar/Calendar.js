import React, { PureComponent } from "react";
import dateFns from "date-fns";
import endOfDay from "date-fns/end_of_day"
import isWithinRange from "date-fns/is_within_range";
import Modal from "../Modal/Modal";
import { getReminders } from '../actions';
import Header from './Header';
import Cells from './Cells';
import { connect } from 'react-redux';

import './calendar.css';

export class Calendar extends PureComponent {
  state = {
    reminders: [],
    currentMonth: new Date(),
    selectedDate: undefined,
    showModal: false
  };

  componentDidMount() {
    this.props.getReminders();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.items.length === 0 || this.state.reminders.length == 0) {
      this.setState({ reminders: this.props.items });
    }
  }

  renderHeader() {
    return (
      <Header prevMonth={this.prevMonth} nextMonth={this.nextMonth}
        currentMonth={this.state.currentMonth} />
    );
  }

  renderDays() {
    const dateFormat = "dddd";
    const days = [];
    const startDate = dateFns.startOfWeek(this.state.currentMonth, { weekStartsOn: 1 });
    for (let i = 0; i < 7; i++) {
      days.push(
        <div className="col col-center" key={i}>
          {dateFns.format(dateFns.addDays(startDate, i), dateFormat)}
        </div>
      );
    }
    return <div className="days row">{days}</div>;
  }

  showModal = async (day) => {
    await this.setState({
      selectedDate: dateFns.format(day, "DD/MM/YYYY")[0],
      showModal: !this.state.showModal
    });
  };

  renderCells() {
    const { currentMonth, selectedDate, reminders } = this.state;
    return (
      <Cells currentMonth={currentMonth} selectedDate={selectedDate} reminders={reminders}
        onDateClick={this.onDateClick}></Cells>
    );
  }

  nextMonth = async () => await this.setState({ currentMonth: dateFns.addMonths(this.state.currentMonth, 1) });
  prevMonth = () => this.setState({ currentMonth: dateFns.subMonths(this.state.currentMonth, 1) });

  onDateClick = day => {
    this.showModal(day);
    this.setState({ selectedDate: dateFns.format(day, "DD/MM/YYYY") });
  };

  onClose = () => this.setState({ showModal: false });

  updateReminder = (message) => {
    const { reminders, selectedDate } = this.state;
    const index = reminders.findIndex(item => item.date === selectedDate);
    const updatedData = index === -1 ? [...reminders, { date: selectedDate, message }]
      : reminders.map(obj => obj.date === selectedDate ? ({ date: obj.date, message }) : obj);
    this.setState({ reminders: updatedData, showModal: false })
  }

  onDelete = () => {
    const reminders = this.state.reminders.filter(item => item.date !== this.state.selectedDate);
    this.setState({ reminders, showModal: false })
  }

  render() {
    const { showModal } = this.state;
    const selectedDate = (this.state.reminders.filter(obj => obj.date === this.state.selectedDate) || []);
    const isDateExist = selectedDate.length > 0;

    return (
      <div className="calendar">
        {this.renderHeader()}
        {this.renderDays()}
        {this.renderCells()}
        <Modal isDateExist={isDateExist} show={showModal} updateReminder={this.updateReminder} onDelete={this.onDelete}
          onClose={this.onClose} message={selectedDate[0] && selectedDate[0].message}></Modal>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    items: state.reminders
  };
};


export default connect(mapStateToProps, { getReminders })(Calendar);

