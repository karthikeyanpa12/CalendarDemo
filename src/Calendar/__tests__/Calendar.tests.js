import React from 'react';
import "babel-polyfill";

import { Calendar } from '../Calendar';
import { shallow } from 'enzyme';

const data = [
    {
        "date": "09/08/2019",
        "message": "Birthday"
    },
    {
        "date": "10/08/2019",
        "message": "New Year"
    }
];

describe('calendar', () => {

    beforeEach(() => {
        jest.restoreAllMocks();
    });

    const tree = shallow(<Calendar getReminders={jest.fn()} items={data} />);
    const instance = tree.instance();
    const currentMonth = new Date("2019-05-10");
    instance.setState({ currentMonth });

    test('renders correctly', async () => {
        //Calendar Snapshot
        expect(tree).toMatchSnapshot();

        //header snapshot
        expect(instance.renderHeader()).toMatchSnapshot();

        //Cells snapshost
        expect(instance.renderCells()).toMatchSnapshot();
    });

    test('renders nextMonth', async () => {
        await instance.nextMonth();
        const month = new Date(instance.state.currentMonth)
        expect(month.getMonth() + 1).toEqual(6);
    });

    test('renders prevMonth', async () => {
        await instance.prevMonth();
        const month = new Date(instance.state.currentMonth)
        expect(month.getMonth()).toEqual(4);
    });

    test('selected Date', async () => {
        const currentDate = new Date("2019-05-10");
        await instance.onDateClick(currentDate);
        const date = new Date(instance.state.selectedDate)
        expect(date.getDate()).toEqual(5);
    });

    test('add Reminder', async () => {
        await instance.updateReminder("welcome");
        const length = instance.state.reminders.length;
        expect(length).toEqual(3);
        expect(instance.state.reminders[length - 1].message).toEqual("welcome");
    });

    test('update Reminder', async () => {
        await instance.updateReminder("welcome123");
        const length = instance.state.reminders.length;
        expect(length).toEqual(3);
        expect(instance.state.reminders[length - 1].message).toEqual("welcome123");
    });

    test('delete Reminder', async () => {
        const currentDate = new Date("2019-05-10");
        await instance.onDelete(currentDate);
        const length = instance.state.reminders.length;
        expect(length).toEqual(2);
    });
});