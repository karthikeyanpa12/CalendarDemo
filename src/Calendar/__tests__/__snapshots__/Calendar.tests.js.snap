// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendar renders correctly 1`] = `ShallowWrapper {}`;

exports[`calendar renders correctly 2`] = `
<header
  currentMonth={2019-05-10T00:00:00.000Z}
  nextMonth={[Function]}
  prevMonth={[Function]}
/>
`;

exports[`calendar renders correctly 3`] = `
<Cells
  currentMonth={2019-05-10T00:00:00.000Z}
  onDateClick={[Function]}
  reminders={
    Array [
      Object {
        "date": "09/08/2019",
        "message": "Birthday",
      },
      Object {
        "date": "10/08/2019",
        "message": "New Year",
      },
    ]
  }
/>
`;
